### PR TITLE
Referenced By View

### PR DESCRIPTION
**Pull request by schofida**
Here you go. It may require some refactoring down the road as there is a lot of duplicated code amongst the views. However, everything appears to be working.

2 Open items (IMO not show stoppers but something to keep in mind):
- So I created 3 new columns for the Referenced By View to match Tes5Edit. These are not sortable and that functionality appears to be going back to the xelib.
- Maybe you can figure this out as maybe something's not being released properly. But I can drag a TreeView item onto the Referenced By View the first time but once it has a record associated with it, I cannot do it anymore. Dbl Click and Enter key press work fine.

### CODE REVIEW

**Comment by matortheeternal** [comment]
There's a lot to review here.  I'll go over tomorrow!

File: app/partials/referencedByView.html [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+</div>
+<div class="tree-nodes" tabindex="0" ng-keydown="onTreeKeyDown($event)" ng-mousedown="onTreeMouseDown($event)" theme-scrollbar-fix vs-repeat>
+    <div ng-repeat="node in tree" ng-mousedown="onNodeMouseDown($event, node)" ng-dblclick="onNodeDoubleClick($event, node)" ng-drag="onNodeDrag(node)" ng-class="{'selected' : node.selected}" class="node {{node.class}}" ng-init="node.has_data || getNodeData(node)">
+        <span ng-repeat="value in node.column_values track by $index" title="{{::value}}" class="column column-{{::$index}}">
+            {{::value}}
+        </span>
+    </div>
+</div>
+<div ng-if="!record" class="message-container" ng-drop="onDrop()" drag-over="onDragOver()">
+    <div ng-if="view.linkedTreeView" class="pane-message">
```
Comment (matortheeternal) [inline app/partials/referencedByView.html]: Originally my idea for this view was not to link it to the tree view, but to link it to the record view, which would make it match xEdit's functionality more closely.  But either way could work.  Some though will have to go into which UX is better.

File: app/partials/referencedByView.html [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<record-address-bar ng-if="showAddressBar"></record-address-bar>
+<!-- TODO: search bar disabled for alpha because not coded yet -->
+<search-bar ng-if="false" ng-controller="recordSearchController"></search-bar>
+<div ng-if="record" class="column-wrapper">
+    <div class="tree-columns">
+        <span ng-repeat-start="column in columns" class="column column-{{::$index}}" ng-class="{'sortable' : column.canSort}" ng-click="toggleSort(column)">
+            {{::column.label}}
+            <span class="sort-arrow" ng-if="column.label === sort.column" title="Elements are sorted by this column in {{sort.reverse ? 'ascending' : 'descending'}} order.">
```
Comment (matortheeternal) [inline app/partials/referencedByView.html]: If sorting isn't working then this shouldn't be here.  Should we get sorting working?

I think in this case we can do the sorting in JavaScript directly, or just not have sorting at all.  I personally don't see much value in sorting the Referenced By view.

File: app/partials/referencedByView.html [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+                <i class="fa fa-arrow-{{sort.reverse ? 'up' : 'down'}}"></i>
+            </span>
+        </span>
+        <split-bar ng-repeat-end ng-if="!$last" offset="16" resize-callback="columnResized"></split-bar>
+        <div class="column-action edit-columns" ng-click="openColumnsModal()" title="Edit columns"></div>
```
Comment (matortheeternal) [inline app/partials/referencedByView.html]: Is there any need to support custom columns in this view?  Maybe?  Can we think of a potentially useful custom column?

File: src/javascripts/Services/recordViewService.js [hunk 4]
```diff
@@ -331,5 +331,11 @@ ngapp.service('recordViewService', function($timeout, layoutService, settingsSer
             scope.view.linkTo(treeView);
             treeView.linkTo(scope.view);
         };
+
+        scope.syncWithLinkedViews = function(record) {
```
Comment (matortheeternal) [inline src/javascripts/Services/recordViewService.js]: This code looks like it's here to sync the record view with the referenced by view, but it's kind of messy.  Traversing through the tree view to the tree view's linked record view.  We shouldn't do it this way.  If we want to link the record view to the referenced by view we should link the referenced by view *directly* to the record view.

File: src/javascripts/Services/referencedByViewColumnsService.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
+ngapp.service('referencedByViewColumnsService', function(settingsService) {
+    let service = this,
+        settings = settingsService.settings;
+
+    let longNameColumn = {
+        label: 'Record',
+        canSort: true,
+        getData: function(node, xelib) {
+            if (node.element_type === xelib.etMainRecord) {
```
Comment (matortheeternal) [inline src/javascripts/Services/referencedByViewColumnsService.js]: I'm pretty sure the referencedByView is never going to display any nodes other than main record nodes, so this kind of conditional isn't really necessary, is it?

File: app/partials/referencedByView.html [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    </div>
+</div>
+<div class="tree-nodes" tabindex="0" ng-keydown="onTreeKeyDown($event)" ng-mousedown="onTreeMouseDown($event)" theme-scrollbar-fix vs-repeat>
+    <div ng-repeat="node in tree" ng-mousedown="onNodeMouseDown($event, node)" ng-dblclick="onNodeDoubleClick($event, node)" ng-drag="onNodeDrag(node)" ng-class="{'selected' : node.selected}" class="node {{node.class}}" ng-init="node.has_data || getNodeData(node)">
```
Comment (matortheeternal) [inline app/partials/referencedByView.html]: creating a whole "tree view" in the referenced by view seems kind of overkill to me.  If we don't have nodes that expand or multiple types of nodes we should be able to use a simpler implementation.  What I'm suggesting won't change the HTML much, but it will change the JavaScript for sure.

File: src/javascripts/Services/referencedByViewColumnsService.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
+ngapp.service('referencedByViewColumnsService', function(settingsService) {
+    let service = this,
+        settings = settingsService.settings;
+
+    let longNameColumn = {
```
Comment (matortheeternal) [inline src/javascripts/Services/referencedByViewColumnsService.js]: It's interesting that xEdit displays LongName in a column, signature in another column, and file in another.  I feel like we could maybe display different columns by default.

Perhaps something like: Signature, FormID, EditorID, Name, File.  Where Name would display similar to how it displays on the tree view (`FullName` or `PlacementName`).  That said, using such a similar column loadout could make it confusing with the tree view, so we should experiment.

Because we don't have native sorting OR node expansion on this view we should be able to make every column optional from the edit columns modal.

File: app/partials/referencedByView.html (hunk 3 above)
Comment (matortheeternal) [reply app/partials/referencedByView.html]: Per my later comment, I do think having custom columns is sensible here.

File: src/javascripts/Services/referencedByViewColumnsService.js [hunk 8]
```diff
@@ -0,0 +1,114 @@
+ngapp.service('referencedByViewColumnsService', function(settingsService) {
```
Comment (matortheeternal) [inline src/javascripts/Services/referencedByViewColumnsService.js]: definitely a lot of repetition here with the `columnsService`.  Perhaps we can figure out a way to extend the `columnsService` to store columns generically?  We can move the column prototypes to a `columnsFactory` and then adjust the functions to work with any view.

File: src/javascripts/Services/referencedByViewService.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+        scope.buildTree = function() {
+            xelib.SetSortMode(scope.sort.column, scope.sort.reverse);
+            scope.tree = xelib.GetReferencedBy(scope.record).map(function(handle) {
+                return scope.buildNode(handle);
+            });
+        };
+
+        // PUBLIC
+        scope.onDragOver = function() {
```
Comment (matortheeternal) [inline src/javascripts/Services/referencedByViewService.js]: if we end up linking this to the record view we won't need drag and drop

File: src/javascripts/Views/editColumnsModal.js [hunk 10]
```diff
@@ -39,15 +39,16 @@ ngapp.controller('editColumnsModalController', function($scope, xelibService, co
 
     $scope.close = function() {
         $scope.allColumns.forEach(function(column) {
-            if (column.custom) columnsService.buildDataFunction(column);
+            if (column.custom) $scope.columnsService.buildDataFunction(column);
         });
         $scope.$root.$broadcast('rebuildColumns');
-        columnsService.saveColumns();
+        $scope.columnsService.saveColumns();
```
Comment (matortheeternal) [inline src/javascripts/Views/editColumnsModal.js]: passing a service through modalOptions is kind of gross.  We definitely need to use a single generic `columnsService`

File: src/javascripts/Views/recordView.js [hunk 11]
```diff
@@ -24,7 +24,11 @@ ngapp.controller('recordViewController', function($scope, $element, $timeout, ht
 
     $scope.toggleAddressBar = function(visible) {
         $scope.showAddressBar = visible;
-        visible ? $timeout($scope.focusAddressInput, 50) : $scope.treeElement.focus();
```
Comment (matortheeternal) [inline src/javascripts/Views/recordView.js]: does this fix some kind of bug with the record view?  not sure about the motivation for this change (seems outside of the scope of the PR)?

File: src/javascripts/Views/referencedByView.js [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    // inherited functions
+    treeService.buildFunctions($scope, $element);
+    treeViewService.buildFunctions($scope, $element);
+    referencedByViewService.buildFunctions($scope);
+    treeColumnService.buildFunctions($scope, '.referenced-by-view', true);
+    nodeSelectionService.buildFunctions($scope, false);
+    hotkeyService.buildOnKeyDown($scope, 'onTreeKeyDown', 'referencedByView');
+
+    $scope.open = function(node) {
+        const record = node.handle ? node.handle : node;
```
Comment (matortheeternal) [inline src/javascripts/Views/referencedByView.js]: what exactly is this line for?

File: src/javascripts/Views/referencedByView.js [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        $scope.syncWithLinkedViews(record);
+    };
+
+    // scope functions
+    $scope.showContextMenu = function(e) {
+        contextMenuService.showContextMenu($scope, e);
+    };
+
+    $scope.toggleAddressBar = function(visible) {
+        $scope.showAddressBar = visible;
```
Comment (matortheeternal) [inline src/javascripts/Views/referencedByView.js]: do we really need to use an address bar on the Referenced By view?  I feel like we don't.  Especially if we link it to the record view.

File: src/javascripts/Views/referencedByView.js [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    };
+
+    $scope.openColumnsModal = function() {
+        $scope.$emit('openModal', 'editColumns', {
+            allColumns: $scope.allColumns,
+            columnsService: referencedByViewColumnsService
+        });
+    };
+
+    $scope.onScroll = function(e) {
```
Comment (matortheeternal) [inline src/javascripts/Views/referencedByView.js]: this code was used on the record view because the columns on the record view may often exceed the width of the record view because it displays override records horizontally.  This code shouldn't be necessary for the referenced by view, which should display columns in a similar fashion to the tree view (fixed width, no horizontal scroll allowed).

I do notice that xEdit supports horizontal scroll on both the referenced by view and the tree view, but if we're going to add that we should do it as a separate PR so we can be consistent.  If the referenced by view is going to support horizontal scroll the tree view should as well.  (the logic related to their columns will have to change a bit, it's too big a change to do in this PR)

File: src/javascripts/Views/referencedByView.js [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+        e.stopImmediatePropagation();
+    };
+
+    // initialization
+    $scope.$on('setRecord', function(e, record) {
+        $scope.open(record);
+        e.stopPropagation();
+    });
+    $scope.$on('reloadGUI', function() {
+        if (!$scope.record) return;
```
Comment (matortheeternal) [inline src/javascripts/Views/referencedByView.js]: if we link this to the record view we should be able to remove all of this logic, as the record view already does it.  The record view will simply set the referenced by view's `$scope.record` from its `$scope.record`.

File: src/javascripts/Services/referencedByViewService.js (hunk 9 above)
Comment (matortheeternal) [reply src/javascripts/Services/referencedByViewService.js]: (well, we could keep drag and drop so the referenced by view can still display records even if it isn't linked to a record view)  :thinking:

File: app/partials/referencedByView.html (hunk 2 above)
Comment (matortheeternal) [reply app/partials/referencedByView.html]: Actually, sorting by signature is kind of useful...  So we should support sorting via JS somehow.

Comment (matortheeternal) [reply app/partials/referencedByView.html]: Sorting from JS will require loading all of the data we're sorting on for the `nodes`.  This shouldn't take a noticeable amount of time unless we're dealing with 500+ nodes, which I don't think is the case for most records.  Definitely something that will need to be tested though.

File: src/javascripts/Views/referencedByView.js (hunk 15 above)
Comment (matortheeternal) [reply src/javascripts/Views/referencedByView.js]: actually, we'll need to keep some of it if we want to allow drag and drop from the tree view.  :thinking:

**Comment by schofida** [comment]
I think that I have made the requested changes. Please let me know if you need anything else. Also, let me know if you see the latest. I am not sure if I got the pull request right.

**Comment by matortheeternal** [comment]
Will take a look soon.

File: src/javascripts/Factories/referencedByViewFactory.js [hunk 16]
```diff
@@ -9,12 +9,16 @@ ngapp.service('referencedByViewFactory', function(viewFactory) {
         let scope = this.scope;
         scope.tree && factory.releaseTree(scope.tree);
         viewFactory.unlink(this.linkedTreeView, 'linkedReferencedByView');
+        viewFactory.unlink(this.linkedRecordView, 'linkedReferencedByView');
     };
 
     this.linkTo = function(view) {
         if (view.class === 'tree-view') {
```
Comment (matortheeternal) [inline src/javascripts/Factories/referencedByViewFactory.js]: The `referencedByView` should not be able to link to a tree view anymore, if we're linking it to a record view.  a view can't have two links in the same direction, it won't make sense.

Here's a diagram of how view linking should work:

File: src/javascripts/Factories/treeViewFactory.js [hunk 17]
```diff
@@ -27,8 +27,7 @@ ngapp.service('treeViewFactory', function(viewFactory) {
         if (view.class === 'record-view') {
             view.linkedTreeView = this;
             this.linkedRecordView = view;
-        }
-        if (view.class === 'referenced-by-view') {
+        } else if (view.class === 'referenced-by-view') {
```
Comment (matortheeternal) [inline src/javascripts/Factories/treeViewFactory.js]: this code can also be removed, the tree view does not need to be able to link to the referenced by view.

File: src/javascripts/Services/columnsService.js [hunk 18]
```diff
@@ -96,26 +63,22 @@ ngapp.service('columnsService', function(settingsService, nodeHelpers) {
...
-            label: 'Name'
-        }]
-    };
-
     this.loadColumns = function() {
-        let data = fh.loadJsonFile('columns.json') || defaultColumnsConfig;
-        service.columns = [formIDColumn, editorIDColumn, nameColumn];
-        data.customColumns.forEach((column) => service.addColumn(column));
-        data.activeColumns.forEach((column) => service.setColumnData(column));
+        let fileData = fh.loadJsonFile('columns.json');
```
Comment (matortheeternal) [inline src/javascripts/Services/columnsService.js]: I'm wondering if it would make sense to have separate JSON files for each view's columns?  Something worth considering.

File: src/javascripts/Services/columnsService.js [hunk 19]
```diff
@@ -96,26 +63,22 @@ ngapp.service('columnsService', function(settingsService, nodeHelpers) {
...
+        columnsFactory.views.forEach(view => {
+            let data = (!fileData || !fileData[view.viewId]) 
+                ? view.defaultColumnsConfig : fileData[view.viewId];
+            service.views[view.viewId] = {};
```
Comment (matortheeternal) [inline src/javascripts/Services/columnsService.js]: This wouldn't be necessary at all if the `columnsFactory` stored views on unique properties, as I previously suggested.

File: src/javascripts/Services/gridService.js [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+// Utilized by all views. Functions for specialized views (e.g. treeView/recordView)
+// will overwrite most of these base functions.
+ngapp.service('gridService', function($timeout, htmlHelpers) {
+    this.buildFunctions = function(scope, element) {
+        // scope functions
+        scope.reload = function() {
+            if (!scope.tree) return;
+            let oldTree = scope.tree;
+            scope.clearSelection(true);
+            scope.buildTree();
```
Comment (matortheeternal) [inline src/javascripts/Services/gridService.js]: should the function really be `buildTree` if we have a grid in this case?  just a thought, might not be worth it to change at the moment.

File: src/javascripts/Services/gridService.js [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+        scope.resolveNodeError = (path, part) => {
+            return new Error(`Failed to resolve node "${part}" in path "${path}"`);
+        };
+
+        scope.onNodeMouseDown = function(e, node) {
+            if (e.button !== 2 || !node.selected) scope.selectNode(e, node);
+            if (e.button === 2) scope.showContextMenu(e);
+        };
+
+        scope.resolveElements = function() {
```
Comment (matortheeternal) [inline src/javascripts/Services/gridService.js]: is this function actually used?  I don't think we need these elements for anything, currently?

File: src/javascripts/Services/gridService.js [hunk 22]
```diff
@@ -0,0 +1,90 @@
+// Base view to show nodes and columns. It's like a tree view but with no depth. 
+// Utilized by all views. Functions for specialized views (e.g. treeView/recordView)
```
Comment (matortheeternal) [inline src/javascripts/Services/gridService.js]: oh, interesting.  I wasn't expecting this, but it may work out.  will have to pay close attention then.

File: src/javascripts/Services/gridService.js [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+            if (width > 100) {
+                let defaultWidth = Math.floor(100 / scope.columns.length) + '%';
+                scope.columns.slice(0, -1).forEach(column => {
+                    column.width = defaultWidth
+                });
+            }
+            scope.resizeColumns();
+        };
+
+        scope.resolveNode = function(path) {
```
Comment (matortheeternal) [inline src/javascripts/Services/gridService.js]: node resolution like this isn't relevant for a grid view.

File: src/javascripts/Services/gridService.js [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+            return node;
+        };
+
+        scope.buildColumnValues = function(node) {
+            node.column_values = scope.columns.map(function(column) {
+                try {
+                    return column.getData(node, xelib);
+                } catch (x) {
+                    console.log(x);
+                    return { value: '' };
```
Comment (matortheeternal) [inline src/javascripts/Services/gridService.js]: this actually is a bug in the code, and should just return and empty string instead.  (was a bug in the code before your refactor which I hadn't fixed)

File: src/javascripts/Services/gridService.js [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+                }
+            }).trimFalsy();
+        };
+
+        scope.getNodeData = function(node) {
+            node.has_data = true;
+            node.fid = xelib.GetFormID(node.handle);
```
Comment (matortheeternal) [inline src/javascripts/Services/gridService.js]: this code is alarming to me.  this code is assuming the node has a `handle` property, and the handle is for an etMainRecord.  That is not going to be the case most of the time for the Record View or Tree View.

File: src/javascripts/Services/gridService.js [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+            scope.buildColumnValues(node);
+        };
+
+        scope.getNodeClass = function(node) {
+            node.class = ' ';
```
Comment (matortheeternal) [inline src/javascripts/Services/gridService.js]: I don't really understand the purpose of this function.  Is this function getting overridden by some other functions in certain cases?

File: src/javascripts/Services/gridService.js [hunk 27]
```diff
@@ -0,0 +1,90 @@
...
+        };
+
+        scope.buildNode = function(handle) {
+            return {
```
Comment (matortheeternal) [inline src/javascripts/Services/gridService.js]: I also do not understand this function.  it seems very simplistic.

I do know I had a buildNode function for the tree view and the record view, and I can see how it might be this simple for a grid view, but I'm still curious.

File: src/javascripts/Services/gridService.js (hunk 22 above)
Comment (matortheeternal) [reply src/javascripts/Services/gridService.js]: Ultimately, I'm not sure if this is even a good idea.  While reducing code repetition is a good thing this may be making things too complicated.  I'll finish my review and come back to this, but I'm tempted to say I'd prefer to have some code repetition and make the grid component code more independent of the tree code.

File: src/javascripts/Services/referencedByViewService.js [hunk 28]
```diff
@@ -4,10 +4,10 @@ ngapp.service('referencedByViewService', function($timeout, layoutService, setti
         scope.releaseTree = referencedByViewFactory.releaseTree;
 
         scope.buildTree = function() {
-            xelib.SetSortMode(scope.sort.column, scope.sort.reverse);
             scope.tree = xelib.GetReferencedBy(scope.record).map(function(handle) {
                 return scope.buildNode(handle);
             });
+            scope.sortTree();
```
Comment (matortheeternal) [inline src/javascripts/Services/referencedByViewService.js]: I dislike the fact that the referencedByViewService has functions with `tree` in their names.  We don't have a tree on this view, so it's kind of a misnomer.

File: src/javascripts/Services/referencedByViewService.js [hunk 29]
```diff
@@ -19,10 +19,46 @@ ngapp.service('referencedByViewService', function($timeout, layoutService, setti
...
+        };
+
+        scope.sortTree = function() {
+            const colIndex = scope.allColumns.findIndex(column => {
+                return column.label === scope.sort.column;
+            });
+            if (colIndex === -1) return;
+            scope.tree.sort((a, b) => {
+                if (!a.has_data) scope.getNodeData(a);
```
Comment (matortheeternal) [inline src/javascripts/Services/referencedByViewService.js]: doing this in the comparator is unnecessarily slow.  I'd prefer a function call before this which gets data for all nodes, or even just getting data for all nodes as a part of `buildTree`.  If we do it in `scope.buildTree` we can just do the following:

```js
scope.tree.forEach(scope.getNodeData);
```

the end result is the same and we don't have to check every node's has_data property n * log(n)+ times.

File: src/javascripts/Services/referencedByViewService.js [hunk 30]
```diff
@@ -19,10 +19,46 @@ ngapp.service('referencedByViewService', function($timeout, layoutService, setti
...
+                scope.tree.reverse();
+            }
+        }
+
+        scope.linkToRecordView = function() {
+            let recordView = layoutService.findView(function(view) {
+                return view.class === 'record-view' && !view.linkedReferencedByView;
+            });
+            if (!recordView) {
+                scope.linkToTreeView();
```
Comment (matortheeternal) [inline src/javascripts/Services/referencedByViewService.js]: nope, don't even bother with this.  if there's no record view don't link it to anything.  the user can then either drag and drop records from the tree view **OR** create a record view, which should automatically link itself to the first unlinked referenced by view it finds.

File: src/javascripts/Views/recordView.js [hunk 31]
```diff
@@ -157,11 +156,17 @@ ngapp.controller('recordViewController', function($scope, $element, $timeout, ht
             $scope.buildColumns();
             $scope.buildTree();
             $scope.$broadcast('recordChanged');
+            let referencedByView = $scope.view.linkedReferencedByView;
```
Comment (matortheeternal) [inline src/javascripts/Views/recordView.js]: I would opt to make this a separate function, to be honest.  Something like `syncReferencedByView`.

File: src/javascripts/Views/referencedByView.js [hunk 32]
```diff
@@ -1,48 +1,36 @@
...
 
     $scope.open = function(node) {
-        const record = node.handle ? node.handle : node;
-        $scope.record = record;
-        $scope.syncWithLinkedViews(record);
+        let recordView = $scope.view.linkedRecordView;
+        if (recordView) {
+            recordView.scope.record = xelib.GetElementEx(node.handle, '');
+        } else {
+            $scope.record = xelib.GetElementEx(node.handle, '');
```
Comment (matortheeternal) [inline src/javascripts/Views/referencedByView.js]: is this behavior we want to have?  navigating through references on the referenced by view by opening them?  it feels somewhat strange to me, but maybe it is sensible when it isn't linked to a record view?  :thinking:

Sure, seems fine.

File: src/javascripts/Views/referencedByView.js [hunk 33]
```diff
@@ -55,15 +43,15 @@ ngapp.controller('referencedByViewController', function($scope, $element, $timeo
     };
 
     $scope.handleEnter = function(e) {
-        $scope.open($scope.lastSelectedNode().handle);
+        $scope.open($scope.lastSelectedNode());
         e.stopImmediatePropagation();
     };
 
+    $scope.releaseHandles = function(oldValue) {
```
Comment (matortheeternal) [inline src/javascripts/Views/referencedByView.js]: if this function is just `xelib.Release`, I don't see why it's necessary.

File: src/javascripts/Views/referencedByView.js (hunk 15 above)
Comment (matortheeternal) [reply src/javascripts/Views/referencedByView.js]: :+1:

File: src/javascripts/Views/referencedByView.js (hunk 33 above)
Comment (matortheeternal) [reply src/javascripts/Views/referencedByView.js]: Does the function this was based off of do more?

File: src/javascripts/Views/referencedByView.js [hunk 34]
```diff
@@ -82,23 +70,19 @@ ngapp.controller('referencedByViewController', function($scope, $element, $timeo
...
-        }
+        $scope.view.label = xelib.Name(newValue);
+        $scope.focusedIndex = -1;
+        $scope.buildColumns();
+        $scope.buildTree();
+        $timeout($scope.resolveElements, 100);
     });
 
-    $scope.showAddressBar = true;
+    $scope.showAddressBar = false;
```
Comment (matortheeternal) [inline src/javascripts/Views/referencedByView.js]: I don't think the referenced by view should have an address bar, so just nuke this line of code, the HTML for the address bar on this view, and any other javascript related to opening it.

File: app/partials/referencedByView.html (hunk 2 above)
Comment (matortheeternal) [reply app/partials/referencedByView.html]: :+1:

File: app/partials/referencedByView.html (hunk 6 above)
Comment (matortheeternal) [reply app/partials/referencedByView.html]: I would like to not use "tree" in classes, functions, or scope variables for this view, because it's not a tree.

File: src/javascripts/Factories/columnsFactory.js [hunk 35]
```diff
@@ -0,0 +1,77 @@
...
+                label: 'Name'
+            }]
+        }
+    }, {
+        viewId: 'referencedByView',
+        defaultColumns: [],
+        defaultColumnsConfig: {
+            customColumns: [{
+                label: 'Record',
+                getDataCode: 'return xelib.LongName(node.handle);'
```
[the remark this answers is not in the document]
Comment (schofida) [reply src/javascripts/Factories/columnsFactory.js]: That is correct. The reason behind adding this is to make all referenced by view columns optional and able to be deleted per your last change review. I am not sure what you mean by a key. You mean adding another property?

File: src/javascripts/Services/columnsService.js (hunk 19 above)
Comment (schofida) [reply src/javascripts/Services/columnsService.js]: Yeah, this line is kinda ugly. But this way, the views can be dynamic. Do we plan on adding that many more views?

File: src/javascripts/Services/referencedByViewService.js (hunk 29 above)
Comment (schofida) [reply src/javascripts/Services/referencedByViewService.js]: Good idea :)

File: src/javascripts/Services/columnsService.js (hunk 19 above)
Comment (matortheeternal) [reply src/javascripts/Services/columnsService.js]: The views can still be dynamic with explicit keys in the `columnsFactory`.  Just do:

```js
this.getColumnsForView(viewName) {
  return columnsFactory[viewName].columns;
}
```

File: src/javascripts/Services/referencedByViewService.js (hunk 29 above)
Comment (matortheeternal) [reply src/javascripts/Services/referencedByViewService.js]: Furthermore, if we build the data all at once when the tree initializes (which we have to do if we want it to be sorted on the JS-side) there's no reason why we need to have `ng-init="node.has_data || getNodeData(node)"` in the HTML.

File: src/javascripts/Factories/columnsFactory.js (hunk 35 above)
[the remark this answers is not in the document]
Comment (matortheeternal) [reply src/javascripts/Factories/columnsFactory.js]: Yeah, I mean we would implement them as hardcoded columns but add a key to them indicating they can be disabled.  E.g.

```js
let recordColumn = {
  label: 'Record',
  canSort: true,
  canBeDisabled: true,
  getData: function(node, xelib) {
    return xelib.LongName(node.handle);
  }
};
```

Then we'd adjust the implementation in the editColumnsModal so columns can be disabled when `canBeDisabled` is truthy and when they are a custom column.

[the remark this answers is not in the document]
Comment (schofida) [reply src/javascripts/Factories/columnsFactory.js]: Do you think that I should worry about this property being persisted in the columns json file?

[the remark this answers is not in the document]
Comment (schofida) [reply src/javascripts/Factories/columnsFactory.js]: Sorry, dumb question. Nevermind.

**Comment by schofida** [comment]
Shoot, I forgot to merge with master before pushing my changes. Let me know if I need to fix anything. Sorry :(

Other than that, I think I finally got it now. Let me know if you see anything else.

**Comment by matortheeternal** [comment]
Looks great!  Good work.  :+1:
